src/services/balancer/contracts/contracts/vault.ts

```diff
@@ -27,88 +38,204 @@ export default class Vault {
+    const validTokens = Object.keys(tokens).filter(
+      address => address !== poolAddress
+    );
+    tokens = pick(tokens, validTokens);
+
+    const normalizedWeights = this.normalizeWeights(
+      data?.weights || [],
+      type,
+      tokens
+    );
+
+    poolData.tokens = this.formatPoolTokens(
+      data.poolTokens,
+      tokens,
+      normalizedWeights,
+      poolAddress
+    );
+
+    poolData.amp = '0';
     if (data?.amp) {
-      amp = data.amp.value.div(data.amp.precision);
+      poolData.amp = data.amp.value.div(data.amp.precision).toString();
     }
 
-    let swapEnabled = true;
-    if (Object.keys(data).includes('swapEnabled')) {
-      swapEnabled = data.swapEnabled;
+    poolData.swapEnabled = true;
+    if (data.swapEnabled !== undefined) {
+      poolData.swapEnabled = data.swapEnabled;
     }
 
-    return {
-      tokens: _tokens,
-      totalSupply: formatUnits(data.totalSupply, data.decimals),
-      decimals: data.decimals,
-      swapFee: formatUnits(data.swapFee, 18),
-      amp,
-      swapEnabled
-    };
+    if (data?.linearPools) {
+      poolData.linearPools = this.formatLinearPools(data.linearPools);
+    }
+
+    poolData.totalSupply = formatUnits(data.totalSupply, data.decimals);
+    poolData.decimals = data.decimals;
+    poolData.swapFee = formatUnits(data.swapFee, 18);
+
+    console.log('onchain data:', poolData);
+
+    return poolData;
+  }
+
+  private formatPoolTokens(
+    poolTokens: RawPoolTokens,
+    tokenInfo: TokenInfoMap,
+    weights: number[],
+    poolAddress: string
+  ): OnchainTokenDataMap {
+    const tokens = <OnchainTokenDataMap>{};
+
+    poolTokens.tokens.forEach((token, i) => {
+      const tokenBalance = poolTokens.balances[i];
+      const decimals = tokenInfo[token]?.decimals;
+      tokens[token] = {
+        decimals,
+        balance: formatUnits(tokenBalance, decimals),
+        weight: weights[i],
+        symbol: tokenInfo[token]?.symbol,
+        name: tokenInfo[token]?.name,
+        logoURI: tokenInfo[token]?.logoURI
+      };
+    });
+
+    // Remove pre-minted BPT
+    delete tokens[poolAddress];
+
+    return tokens;
+  }
+
+  private formatLinearPools(
+    linearPools: RawLinearPoolDataMap
+  ): LinearPoolDataMap {
+    const _linearPools = <LinearPoolDataMap>{};
+
+    Object.keys(linearPools).forEach(address => {
+      const { id, mainToken, wrappedToken, priceRate, tokenData } = linearPools[
+        address
+      ];
+
+      _linearPools[address] = {
+        id,
+        priceRate: formatUnits(priceRate.toString(), 18),
+        mainToken: {
+          address: getAddress(mainToken.address),
+          index: mainToken.index.toNumber(),
+          balance: tokenData.balances[mainToken.index.toNumber()].toString()
+        },
+        wrappedToken: {
+          address: getAddress(wrappedToken.address),
+          index: wrappedToken.index.toNumber(),
+          balance: tokenData.balances[wrappedToken.index.toNumber()].toString()
+        }
```

Comment:
@evgenyboxer this is the linear pool schema I've got so far we can add to it or adjust it as needed. Note that the balance is scaled and would need to be normalised with the token decimals, I don't think we should do that here since passing in the token info for the sub tokens is a bit tricky.